### PR TITLE
PLFM-5604 google cloud storage client code

### PR DESCRIPTION
**Pull request by nickgros**

### CODE REVIEW

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/DefaultGCPCredentialsProvider.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package org.sagebionetworks.gcp;
+
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.auth.Credentials;
+import com.google.cloud.storage.StorageOptions;
+
+/**
+ * A AWSCredentialsProvider that will attempt to load AWS credentials from the
```
Comment (nickgros) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/DefaultGCPCredentialsProvider.java]: Fix the comment, note the environment var

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGCPCredentialsProviderChain.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ * following order:
+ * <ol>
+ * <li>sage.bionetworks properties from Maven .m2/settings.xml file</li>
+ * <li>sage.bionetworks properties from Java System.properties</i>
+ * <li>the default system credentials</li>
+ * </ol>
+ */
+public class SynapseGCPCredentialsProviderChain implements CredentialsProvider {
+
+	private List<CredentialsProvider> credentialsProviders = new LinkedList<>();
```
Comment (zimingd) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGCPCredentialsProviderChain.java]: pull into constructor

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGCPCredentialsProviderChain.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	static private Logger log = LogManager.getLogger(SynapseGCPCredentialsProviderChain.class);
+
+	private static final SynapseGCPCredentialsProviderChain INSTANCE =
+			new SynapseGCPCredentialsProviderChain(new PropertyProviderImpl(), new DefaultGCPCredentialsProvider());
```
Comment (zimingd) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGCPCredentialsProviderChain.java]: either take all providers in constructor as args or take no args and predefine them in constructor.

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGCPCredentialsProviderChain.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	 * Access the singleton chain.
+	 * @return
+	 */
+	public static SynapseGCPCredentialsProviderChain getInstance() {
+		return INSTANCE;
+	}
+
+	@Override
+	public Credentials getCredentials() {
+		List<String> exceptionMessages = null;
```
Comment (zimingd) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGCPCredentialsProviderChain.java]: instantiate a linked list

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGCPCredentialsProviderChain.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		for (CredentialsProvider provider : credentialsProviders) {
+			try {
+				Credentials credentials = provider.getCredentials();
+				if (credentials != null) {
+					log.debug("Loading credentials from " + provider.toString());
+					return credentials;
+				}
+			} catch (Exception e) {
+				// Ignore any exceptions and move onto the next provider
+				String message = provider + ": " + e.getMessage();
```
Comment (zimingd) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGCPCredentialsProviderChain.java]: rename to providerErrorMessage

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGoogleCloudStorageClient.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	 * @throws IOException if the file cannot be opened
+	 */
+	Blob putObject(String bucket, String key, InputStream inputStream) throws IOException;
+
+	/**
+	 * Deletes the object in Google Cloud Storage
+	 * @param bucketName The bucket containing the object
+	 * @param key The path/filename of the object
+	 */
+	void deleteObject(String bucketName, String key);
```
Comment (zimingd) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGoogleCloudStorageClient.java]: bucket

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGoogleCloudStorageClientImpl.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	@Override
+	public Blob composeObjects(String bucket, String newKey, List<String> partKeys) throws StorageException {
+		if (partKeys.size() > MAX_OBJECTS_IN_COMPOSE)
+			throw new IllegalArgumentException("Cannot compose more than " + MAX_OBJECTS_IN_COMPOSE + " objects in one request");
+		BlobInfo blobInfo = BlobInfo.newBuilder(BlobId.of(bucket, newKey)).build();
+		Storage.ComposeRequest.Builder builder = Storage.ComposeRequest.newBuilder().setTarget(blobInfo);
+		for (String part : partKeys) {
+				builder.addSource(part);
```
Comment (zimingd) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGoogleCloudStorageClientImpl.java]: addSource(partKeys) is supported

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGoogleCloudStorageClientImpl.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+		}
+		return storage.compose(builder.build());
+	}
+
+	@Override
+	public void rename(String bucket, String oldKey, String newKey) throws StorageException {
+		storage.copy(Storage.CopyRequest.newBuilder().setSource(BlobId.of(bucket,oldKey)).setTarget(BlobId.of(bucket,newKey)).build());
```
Comment (zimingd) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGoogleCloudStorageClientImpl.java]: space out onto multiple lines

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseSystemPropertiesGCPCredentialsProvider.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * An AWS credential provider that will look for:
+ * 'org.sagebionetworks.stack.iam.id' and 'org.sagebionetworks.stack.iam.key' on
+ * the Java System.properties.
+ *
+ */
+public class SynapseSystemPropertiesGCPCredentialsProvider extends AbstractSynapseGcpCredentialsProvider {
+
+	Properties properties;
```
Comment (zimingd) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseSystemPropertiesGCPCredentialsProvider.java]: make abstract class have a final Properties field and set it in the abstract class's constructors. 
and call super(properties)

File: lib/stackConfiguration/src/test/java/org/sagebionetworks/SynapseGoogleCloudStorageClientImplUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+		assertThrows(RuntimeException.class, () -> client.deleteObject(BUCKET_NAME, OBJECT_KEY));
+	}
+
+	@Test
+	public void createSignedUrl() {
+		long expirationTime = 50L;
+		HttpMethod method = HttpMethod.GET;
+		client.createSignedUrl(BUCKET_NAME, OBJECT_KEY, expirationTime, method);
+
+		verify(mockStorage).signUrl(eq(OBJECT_BLOB_INFO), eq(expirationTime),
```
Comment (zimingd) [inline lib/stackConfiguration/src/test/java/org/sagebionetworks/SynapseGoogleCloudStorageClientImplUnitTest.java:108]: remove captors if unused

File: lib/stackConfiguration/src/test/java/org/sagebionetworks/SynapseGoogleCloudStorageClientImplUnitTest.java [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+		assertEquals(mockBlob, result);
+		verify(mockStorage).get(OBJECT_BLOB_ID);
+	}
+
+	@Test
+	public void putObject() {
+		byte[] fileContents = "some file content".getBytes(StandardCharsets.UTF_8);
+		InputStream inputStream = new ByteArrayInputStream(fileContents);
+		// We use a stubbed class to verify that the contents are copied properly
+		WriteChannelStub writableByteChannel = new WriteChannelStub();
```
Comment (zimingd) [inline lib/stackConfiguration/src/test/java/org/sagebionetworks/SynapseGoogleCloudStorageClientImplUnitTest.java]: reanme to TestStubWriteChannel

File: lib/stackConfiguration/src/test/java/org/sagebionetworks/SynapseGoogleCloudStorageClientImplUnitTest.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+	public void deleteObjectSuccess() {
+		when(mockStorage.delete(OBJECT_BLOB_ID)).thenReturn(true);
+		client.deleteObject(BUCKET_NAME, OBJECT_KEY);
+		verify(mockStorage).delete(OBJECT_BLOB_ID);
+	}
+
+	@Test
+	public void deleteObjectFailure() {
+		when(mockStorage.delete(OBJECT_BLOB_ID)).thenReturn(false);
+		assertThrows(RuntimeException.class, () -> client.deleteObject(BUCKET_NAME, OBJECT_KEY));
```
Comment (zimingd) [inline lib/stackConfiguration/src/test/java/org/sagebionetworks/SynapseGoogleCloudStorageClientImplUnitTest.java]: change runtime to another more specific exception type. we usually only use runtime exceptions to wrap checked exceptions that we don't expect to occur

File: services/repository/pom.xml [hunk 13]
```diff
@@ -249,13 +249,13 @@
 					<destDir>${project.build.directory}/javadoc/restdocs</destDir>
 					<!-- This name is used to link to the authentication controller -->
 					<additionalJOption>-authControllerName org.sagebionetworks.auth.AuthenticationController</additionalJOption>
-					<additionnalDependencies>
```
Comment (zimingd) [inline services/repository/pom.xml]: this is maven's typo
https://maven.apache.org/components/plugins-archives/maven-javadoc-plugin-2.9.1/examples/additionnal-dependencies.html

Comment (zimingd) [reply services/repository/pom.xml]: as a note this is fixed in later versions but upgrading the version breaks the document generation. not sure what causes it, but fixing the typos is more work than a simple version bump

**Review by zimingd — CHANGES_REQUESTED** [review]

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGoogleCloudCredentialsProviderChain.java [hunk 14]
```diff
@@ -3,55 +3,58 @@
...
  * <ol>
  * <li>sage.bionetworks properties from Maven .m2/settings.xml file</li>
  * <li>sage.bionetworks properties from Java System.properties</i>
  * <li>the default system credentials</li>
  * </ol>
  */
-public class SynapseGCPCredentialsProviderChain implements CredentialsProvider {
+public class SynapseGoogleCloudCredentialsProviderChain implements CredentialsProvider {
 
 	private List<CredentialsProvider> credentialsProviders = new LinkedList<>();
```
Comment (zimingd) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGoogleCloudCredentialsProviderChain.java]: move to constructor

File: lib/stackConfiguration/src/test/java/org/sagebionetworks/gcp/SynapseGoogleCloudCredentialsProviderChainTest.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+		when(mockProvider3.getCredentials()).thenReturn(null);
+
+		// call under test
+		assertThrows(RuntimeException.class, () -> chain.getCredentials());
+
+		verify(mockProvider1).getCredentials();
+		verify(mockProvider2).getCredentials();
+		verify(mockProvider3).getCredentials();
+	}
+
```
Comment (zimingd) [inline lib/stackConfiguration/src/test/java/org/sagebionetworks/gcp/SynapseGoogleCloudCredentialsProviderChainTest.java:103]: test cases for exception handling

Comment (zimingd) [reply lib/stackConfiguration/src/test/java/org/sagebionetworks/gcp/SynapseGoogleCloudCredentialsProviderChainTest.java:103]: test singleton instance w/ assertSame()

File: lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGoogleCloudCredentialsProviderChain.java [hunk 16]
```diff
@@ -3,55 +3,58 @@
...
+							new DefaultGoogleCloudCredentialsProvider()
+					)
+			);
+
 	/**
 	 * Access the singleton chain.
 	 * @return
 	 */
-	public static SynapseGCPCredentialsProviderChain getInstance() {
+	public static SynapseGoogleCloudCredentialsProviderChain getInstance() {
```
Comment (zimingd) [inline lib/stackConfiguration/src/main/java/org/sagebionetworks/gcp/SynapseGoogleCloudCredentialsProviderChain.java]: getDefaultChainInstance()

**Review by zimingd — CHANGES_REQUESTED** [review]

**Review by zimingd — APPROVED** [review]

**Pull request merged**